src/commands/plugins.rs

```diff
@@ -0,0 +1,275 @@
+use anyhow::{anyhow, Result};
+use clap::{Parser, Subcommand};
+use semver::Version;
+use spin_plugins::{
+    install::{ManifestLocation, PluginInfo, PluginInstaller},
+    uninstall::PluginUninstaller,
+    PLUGIN_MANIFESTS_DIRECTORY_NAME,
+};
+use std::path::PathBuf;
+use tracing::log;
+use url::Url;
+
+const SPIN_PLUGINS_REPO: &str = "https://github.com/fermyon/spin-plugins/";
+
+/// Install/uninstall Spin plugins.
+#[derive(Subcommand, Debug)]
+pub enum PluginCommands {
+    /// Install plugin from a manifest.
+    ///
+    /// The binary file and manifest of the plugin is copied to the local Spin
+    /// plugins directory.
+    Install(Install),
+
+    /// Remove a plugin from your installation.
+    Uninstall(Uninstall),
+
+    /// Upgrade one or all plugins.
+    Upgrade(Upgrade),
+}
+
+impl PluginCommands {
+    pub async fn run(self) -> Result<()> {
+        match self {
+            PluginCommands::Install(cmd) => cmd.run().await,
+            PluginCommands::Uninstall(cmd) => cmd.run().await,
+            PluginCommands::Upgrade(cmd) => cmd.run().await,
+        }
+    }
+}
+
+/// Install plugins from remote source
+#[derive(Parser, Debug)]
+pub struct Install {
+    /// Name of Spin plugin.
+    #[clap(
+        name = "PLUGIN_NAME",
+        conflicts_with = "REMOTE_PLUGIN_MANIFEST",
+        conflicts_with = "LOCAL_PLUGIN_MANIFEST",
+        required_unless_present_any = ["REMOTE_PLUGIN_MANIFEST", "LOCAL_PLUGIN_MANIFEST"],
+    )]
+    pub name: Option<String>,
+
+    /// Path to local plugin manifest.
+    #[clap(
+        name = "LOCAL_PLUGIN_MANIFEST",
+        short = 'f',
+        long = "file",
+        conflicts_with = "REMOTE_PLUGIN_MANIFEST",
+        conflicts_with = "PLUGIN_NAME"
+    )]
+    pub local_manifest_src: Option<PathBuf>,
+
+    /// Path to remote plugin manifest.
+    #[clap(
+        name = "REMOTE_PLUGIN_MANIFEST",
+        short = 'u',
+        long = "url",
+        conflicts_with = "LOCAL_PLUGIN_MANIFEST",
+        conflicts_with = "PLUGIN_NAME"
+    )]
+    pub remote_manifest_src: Option<Url>,
+
+    /// Skips prompt to accept the installation of the plugin.
+    #[clap(short = 'y', long = "yes", takes_value = false)]
+    pub yes_to_all: bool,
+
+    /// Specific version of a plugin to be install from the centralized plugins
+    /// repository.
+    #[clap(
+        long = "version",
+        short = 'v',
+        conflicts_with = "REMOTE_PLUGIN_MANIFEST",
+        conflicts_with = "LOCAL_PLUGIN_MANIFEST",
+        requires("PLUGIN_NAME")
+    )]
+    pub version: Option<Version>,
+}
+
+impl Install {
+    pub async fn run(self) -> Result<()> {
+        let manifest_location = match (self.local_manifest_src, self.remote_manifest_src, self.name) {
```

Comment:
Could move this lengthy match out to a `self.manifest_location()` function?

Reply:
Using a `ManifestLocation` enum is really nice.  Much cleaner than carting all those options around.